contributors/design-proposals/auth/bound-service-account-tokens.md

```diff
@@ -0,0 +1,180 @@
+attenuated token for it's own use. The API will enforce required attenuations,
+e.g. audience and time binding.
+
+## Token attenuations
+
+### Audience binding
+
+Tokens issued from this API will be audience bound. Audience of requested tokens
+will be bound by the `aud` claim. The `aud` claim is an array of strings
+(usually URLs) that correspond to the intended audience of the token. A
+recipient of a token is responsible for verifying that it identifies as one of
+the values in the audience claim, and should otherwise reject the token. The
+TokenReview API will support this validation.
+
+### Time binding
+
+Tokens issued from this API will be time bound. Time validity of these tokens
+will be claimed in the following fields:
+
+* `exp`: expiration time
+* `nbf`: not before
+* `iat`: issued at
+
+A recipient of a token is valid at the time that the token is presented, and
+should otherwise reject the token. The TokenReview API will support this
+validation.
+
+### Key binding
+
+Tokens issued from this API will OPTIONALLY (per the behest of the requestor)
+be key bound. The `cnf` claim will carry information that a recipient of a
+token might need to validate that the token presenter possesses a specific key.
+A recipient may require accepted tokens are key bound.
+
+## API Changes
+
+### Add `tokenrequests.authentication.k8s.io`
+
+We will add an imperative API (a la TokenReview) to the
+`authentication.k8s.io` API group:
+
+```golang
+type TokenRequest struct {
+  Spec   TokenRequestSpec
+  Status TokenRequestStatus
+}
+
+type TokenRequestSpec struct {
+  // Audience is the intendend audience of the token.
+  //
+  // see audience binding
+  Audience []string
+
+  // ValidityDuration is the intended duration of validity of the request.
+  //
+  // see time binding
+  ValidityDuration metav1.Duration
+
+  // KeyConfirmation contains information required by a recipient of a token
+  // to verify possession of a key.
+  //
+  // see key binding
+  KeyConfirmation KeyConfirmation
+}
+
+type TokenRequestStatus struct {
+  Token string
+}
+
+type KeyConfirmation struct {
+  // Structure TBD.
+  // This will not be supported in the initial implementation. This proposal
+  // will be updated later on.
+}
+
+type TokenRequestStatus struct {
+  // Token is the token data
+  Token string
+}
+```
+
+This API will be exposed as a subresource under a serviceacccount object. A
+requestor for a token for a specific service account will `POST` a
+`TokenRequest` to the `/token` subresource of that service account object.
+
+### Modify `tokenreviews.authentication.k8s.io`
+
+The TokenReview API will be extended to support passing an additional audience
+field which the service account authenticator will validate.
+
+```golang
+type TokenReviewSpec struct {
+  // Token is the opaque bearer token.
+  Token string
+  // Audience is the identifier that the client identifies as.
+  Audience string
```

Comment:
Would be easy-ish to plumb this through the oidc plugin (if aud != clientid -> reject). Don't know what this would look like for the wehbook authenticator or how we'd roll it out. If a webhook authentication backend didn't consider this field you could get in some dangerous situations, right?

Reply:
passing a token intended for someone other than the apiserver to the apiserver for verification seems a little weird

Reply:
> Don't know what this would look like for the wehbook authenticator or how we'd roll it out. If a webhook authentication backend didn't consider this field you could get in some dangerous situations, right?

Is this any different from webhook authenticator and svcacct id_tokens today?

> passing a token intended for someone other than the apiserver to the apiserver for verification seems a little weird

This is true of all usage of TokenReview API endpoint though, isn't it? In this case, it is less weird since the apiserver is the thing creating the tokens in the first place. An API to correctly and opaquely verify these tokens is valuable for integrations.